developer-workflow/c-api.rst

```diff
@@ -134,6 +146,103 @@ When moving existing tests, feel free to replace ``TestError`` with
+  This happens automatically, based on the name
+  (via :cpy-file:`Doc/tools/extensions/c_annotations.py`).
+
+Despite being “unstable”, there are rules to make sure third-party code can
+use this API reliably:
+
+* Changes and removals can be done in minor releases (3.x.0, including
+  Alphas and Betas for 3.x.0).
+* Adding a new unstable API *for an existing feature* is allowed even after
+  Beta feature freeze, up until the first Release Candidate.
+  Consensus on the `Core Development Discourse <https://discuss.python.org/c/core-dev/23>`_
+  is needed in the Beta period.
+* Backwards-incompatible changes should make existing C callers fail to compile.
+  For example, arguments should be added/removed, or a function should be
+  renamed.
+* When moving an API into or out of the Unstable tier, the old name
+  should continue to be available (but deprecated) until an incompatible
+  change is made. In other words, while we're allowed to break calling code,
+  we shouldn't break it *unnecessarily*.
+
+
+Moving an API from the public tier to Unstable
+----------------------------------------------
+
+* Expose the API under its new name, with the ``PyUnstable_`` prefix.
+* Make the old name an alias (e.g. a ``static inline`` function calling the
+  new function).
+* Deprecate the old name, typically using :c:macro:`Py_DEPRECATED`.
+* Announce the change in the "What's New".
+
+The old name should continue to be available until an incompatible change is
+made. Per Python’s backwards compatibility policy (:pep:`387`),
+this deprecation needs to last at least two releases
+(modulo Steering Council exceptions).
+
+The rules are relaxed for APIs that were introduced in Python versions
+before 3.12, when the official Unstable tier was added.
+You can make an incompatible change (and remove the old name)
+as if the function was already part of the Unstable tier
+for APIs introduced before Python 3.12 that are either:
+
+* Documented to be less stable than default.
+* Named with a leading underscore.
+
+Moving an API from the private tier to unstable
+-----------------------------------------------
+
+* Expose the API under its new name, with the ``PyUnstable_`` prefix.
+* If the old name is documented, or widely used externally,
+  make it an alias and deprecate it (typically with :c:macro:`Py_DEPRECATED`).
+  It should continue to be available until an incompatible change is made,
+  as if it was previously public.
+
+  This applies even to underscored names. Python wasn't always strict with
+  the leading underscore.
+* Announce the change in What's New.
+
+Moving an API from unstable to public
+-------------------------------------
+
+* Expose the API under its new name, without the ``PyUnstable_`` prefix.
+* Make the old ``PyUnstable_*`` name be an alias (e.g. a ``static inline``
+  function calling the new function).
+* Announce the change in What's New.
+
+The old name should continue to be available until an incompatible change is
+made.
```

Comment:
I have no subject matter expert opinion on the changes, of course (though ISTM that removing the alias after a backward-incompatible change to a public API should still be at least mentioned, as it is safer than silently changing behavior) but if you want to accept it, I've made @ncoghlan 's comment from the above into an actual suggestion, on the correct lines, and with some copyediting cleanup to make it more concise and avoid repetition.

```suggestion
The previous unstable name should remain available as long as the new public name is.
If the public name is ever subsequently deprecated, its unstable alias should be too.

While the unstable name could theoretically be removed or deprecated immediately without
violating the expectations set for the unstable API tier, there's also no need to break working
code just because a previously unstable API is now considered ready for a wider audience.
```

Or, with an explicit mention of removal after an incompatible change, as before:

```suggestion
The previous unstable name should remain available as long as the new public name is,
unless an incompatible change is made to the public API that could break users anyway.
If the public name is ever subsequently deprecated, its unstable alias should be too.

While the unstable name could theoretically be removed or deprecated immediately without
violating the expectations set for the unstable API tier, there's also no need to break working
code just because a previously unstable API is now considered ready for a wider audience,
unless it may break or change its behavior anyway due to another incompatible change.
```

Reply:
It's a public API, so incompatible changes are not allowed. This new text should *not* suggest that they are. If we want to make an incompatible change to a public API, we have to add a new API and deprecate the old one instead.

Reply:
@CAM-Gerlach It's basically the entirety of the ["Making incompatible changes" section of PEP 387](https://peps.python.org/pep-0387/#making-incompatible-changes). (Particularly steps 2 and 3.)

> ### Making incompatible changes
> Making an incompatible change is a gradual process performed over several releases:
> 
> 1. Discuss the change. Depending on the degree of incompatibility, this could be on the bug tracker, python-dev, python-list, or the appropriate SIG. A PEP or similar document may be written. Hopefully users of the affected API will pipe up to comment.
> 1. Add a warning. If behavior is changing, the API may gain a new function or method to perform the new behavior; old usage should raise the warning. If an API is being removed, simply warn whenever it is entered. `DeprecationWarning` is the usual warning category to use, but `PendingDeprecationWarning` may be used in special cases where the old and new versions of the API will coexist for many releases [[2]](https://peps.python.org/pep-0387/#warnings). Compiler warnings are also acceptable. The warning message should include the release the incompatibility is expected to become the default and a link to an issue that users can post feedback to.
> 1. Wait for the warning to appear in at least two minor Python versions of the same major version, or one minor version in an older major version (e.g. for a warning in Python 3.10, you either wait until at least Python 3.12 or Python 4.0 to make the change). It’s fine to wait more than two releases.
> [...]

Reply:
Well, its certainly not "the entirety of the section"; it's the very specific point above I'm after. On careful inspection I do see what specifically implies this, though it is a bit buried and took multiple reads to find:

> If behavior is changing, the API may gain a new function or method to perform the new behavior; old usage should raise the warning. 

The language here really should be more clear and explicit, as the "may"s and "should"s used here certainly don't send nearly the same message as "incompatible changes are not allowed" and "If we want to make an incompatible change to a public API, we have to add a new API and deprecate the old one instead." Also, "old usage" is somehwt unclear terminology. For that, it should instead state something like

> If an incompatible behavior change to an existing API is needed, a new function or method MUST be added and the old construct deprecated, so existing uses don't silently break.

To note, that isn't always true at least for Python APIs; I can recall several instances (e.g. regex group identifiers) where incompatible changes were made to existing functions for various reasons.

Reply:
Actually, I think the alias should be removed as soon as the public API is deprecated. If you signed up for updating your code to react to implementation changes, you should update it when the API is no longer favoured.

```suggestion
The old name should remain available until the
new public name is deprecated or removed.
```

This is technically a change in the PEP, I'll need to run it by the SC.

Reply:
>  I'll need to run it by the SC.

(I plan to do this after merging, and revert if necessary)